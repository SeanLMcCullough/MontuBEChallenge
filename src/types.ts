import type { AxiosRequestConfig } from 'axios'

// #region GeoSearchEngine

export interface GeoSearchEngine {
  getAutoCompleteDetails: (query: string, options?: GeoSearch) => Promise<GeoSearchResults>
}

export interface GeoSearchEngineConfiguration {
  defaults?: GeoSearch
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface GeoSearch {}

export interface GeoSearchResults {
  results: GeoSearchResult[]
}

export interface GeoSearchResult {
  placeId: string
  address: GeoSearchResultAddress
  position: GeoSearchPoint
}

export interface GeoSearchResultAddress {
  country: string | null
  countryCode: string | null
  countryCodeISO3: string | null
  countrySecondarySubdivision: string | null
  countrySubdivision: string | null
  countrySubdivisionCode: string | null
  countrySubdivisionName: string | null
  countryTertiarySubdivision: string | null
  extendedPostalCode: string | null
  freeformAddress: string | null
  localName: string | null
  municipality: string | null
  municipalitySecondarySubdivision: string | null
  municipalitySubdivision: string | null
  neighbourhood: string | null
  postalCode: string | null
  streetName: string | null
  streetNumber: string | null
}

export interface GeoSearchPoint {
  lat: number
  lon: number
}

// #endregion GeoSearchEngine

// #region TomTomGeoSearchEngine

export interface TomTomGeoSearchEngineConfiguration extends GeoSearchEngineConfiguration {
  apiKey: string
  apiVersion?: string
  axiosConfig?: AxiosRequestConfig
  baseUrl?: string
  defaults?: TomTomGeoSearchOptions
}

export interface TomTomGeoSearchOptions extends GeoSearch {
  boundingBox?: TomTomBoundingBoxOptions // Bounding box to constrain the search results.
  brandSet?: string // List of brand names which could be used to restrict the result to Points Of Interest of specific brands.
  categorySet?: string // List of categories which could be used to restrict the result to the Points Of Interest of specific categories. The list of categories can be discovered using the POI Categories endpoint.
  connectorSet?: string // List of connector types which could be used to restrict the result to the Points Of Interest of type Electric Vehicle Station supporting specific connector types. See: List of supported connector types
  countrySet?: string // Country codes in ISO 3166-1 alpha-2 or alpha-3 code formats (e.g., FR,ES or FRA,ESP). This will limit the search to the specified countries. The choice of view may restrict which countries are available.
  entityTypeSet?: TomTomEntityType | TomTomEntityType[] // A comma-separated list of entity types which can be used to restrict the result to the Geography result of a specific entity type. If entityTypeSet is specified, only a Geography result with a proper entity type will be returned.
  extendedPostalCodesFor?: string // Indexes for which extended postal codes should be included in the results.
  fuelSet?: string // List of fuel types which could be used to restrict the result to the Points Of Interest of specific fuels. If fuelSet is specified, the query can remain empty. Only POIs with a proper fuel type will be returned.
  geoBias?: string // Location bias for search, which can be specified in different shapes. Location bias is used to communicate a location preference to the search engine - for example, to bias search results to the user's viewport. Currently two types are supported: point and rectangle.
  idxSet?: string // List of indexes which should be utilized for the search.
  language?: string // Language in which search results should be returned. It should be one of the TomTom supported IETF language tags , case insensitive. When data in the specified language is not available for a specific field or the language is not specified, the language best matched with your query is used.
  latitude?: number // Latitude, e.g., lat=37.337 lat,lon where results should be biased. Note: supplying a lat/lon without a radius will only bias the search results to that area.
  limit?: number
  longitude?: number // Longitude, e.g., lon=-121.89 lat,lon where results should be biased. Note: supplying a lat/lon without a radius will only bias the search results to that area.
  mapcodes?: string // Enables the return of a comma-separated mapcodes list. Can also filter the response to only show selected mapcode types. See mapcodes in the response.
  maxFuzzyLevel?: number // Maximum fuzzyness level to be used.
  maxPowerKW?: number // An optional parameter which could be used to restrict the result to the Points Of Interest of type Electric Vehicle Station supporting at least one connector with a specific maximum value of power in kilowatts (closed interval - with that value).
  minFuzzyLevel?: number // Minimum fuzzyness level to be used.
  minPowerKW?: number // An optional parameter which could be used to restrict the result to the Points Of Interest of type Electric Vehicle Station supporting at least one connector with a specific minimal value of power in kilowatts (closed interval - with that value).
  offset?: number
  openingHours?: string // List of opening hours for a POI (Points of Interest).
  radius?: number // If radius and position are set the results will be constrained to the defined area. The radius parameter is specified in meters. Valid radius values are numbers greater than 0. Supplying values equal to or lesser than 0 causes the parameter to be ignored.
  relatedPois?: TomTomRelatedPoi // An optional parameter that provides the possibility to return related Points Of Interest.
  timeZone?: string // Used to indicate the mode in which the timeZone object should be returned.
  typeahead?: boolean
  vehicleTypeSet?: TomTomVehicleType | TomTomVehicleType[] // List of vehicle types that could be used to restrict the result to the Points Of Interest of specific vehicles. If vehicleTypeSet is specified, the query can remain empty. Only POIs with a proper vehicle type will be returned.
  view?: string // Geopolitical View. The context used to resolve the handling of disputed territories. Views include Unified , along with AR , IL , IN , MA , PK , RU , TR , and CN which are respectively tailored for Argentina, Israel, India, Morocco, Pakistan, Russia, Turkey, and China.
}

export interface TomTomBoundingBoxOptions {
  topLeft?: string // Top-left position of the bounding box. Note: the btmRight parameter should always go along with the topLeft parameter as they both define the bounding box.
  bottomRight?: string // Bottom-right position of the bounding box. Note: The btmRight parameter should always go along with the topLeft parameter as they both define the bounding box.
}

export type TomTomVehicleType = 'Car' | 'Truck'

export type TomTomMapCode = 'Local' | 'International' | 'Alternative'

export type TomTomRelatedPoi = 'off' | 'child' | 'parent' | 'all'

export type TomTomEntityType =
  | 'Country'
  | 'CountrySubdivision'
  | 'CountrySecondarySubdivision'
  | 'CountryTertiarySubdivision'
  | 'Municipality'
  | 'MunicipalitySubdivision'
  | 'MunicipalitySecondarySubdivision'
  | 'Neighbourhood'
  | 'PostalCodeArea'

export interface TomTomGeoSearchResponse {
  summary: TomTomGeoSearchSummary
  results: TomTomGeoSearchResult[]
}

export interface TomTomGeoSearchSummary {
  query: string
  queryType: string
  queryTime: number
  numResults: number
  offset: number
  totalResults: number
  fuzzyLevel: number
  queryIntent: TomTomGeoSearchQueryIntent[]
}

export interface TomTomGeoSearchQueryIntent {
  type: string
  details: object
}

export interface TomTomGeoSearchResult {
  type: string
  id: string
  score: number
  info: string
  poi: TomTomGeoSearchPoi
  address: GeoSearchResultAddress
  position: GeoSearchPoint
  viewport: TomTomGeoSearchViewport
  entryPoints: TomTomGeoSearchEntryPoint[]
}

export interface TomTomGeoSearchPoi {
  name: string
  categorySet: TomTomGeoSearchCategory[]
  url: string
  categories: string[]
  classifications: TomTomGeoSearchClassification[]
}

export interface TomTomGeoSearchCategory {
  id: number
}

export interface TomTomGeoSearchClassification {
  code: string
  names: TomTomGeoSearchClassificationName[]
}

export interface TomTomGeoSearchClassificationName {
  nameLocale: string
  name: string
}

export interface TomTomGeoSearchViewport {
  topLeftPoint: GeoSearchPoint
  btmRightPoint: GeoSearchPoint
}

export interface TomTomGeoSearchEntryPoint {
  type: string
  position: GeoSearchPoint
}

// #endregion TomTomGeoSearchEngine
