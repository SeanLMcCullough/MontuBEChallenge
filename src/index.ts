export { default as TomTomGeoSearchEngine } from './tomtom-geo-search-engine'
