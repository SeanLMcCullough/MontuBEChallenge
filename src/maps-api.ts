import axios, { type AxiosResponse } from 'axios'
import type { TomTomGeoSearchResponse, GeoSearchResults } from './types'

// https://developer.tomtom.com/search-api/documentation/search-service/fuzzy-search
export async function getPlaceAutocomplete (
  key: string,
  address: string
): Promise<GeoSearchResults> {
  const autocomplete: AxiosResponse<TomTomGeoSearchResponse> = await axios.get(
    `https://api.tomtom.com/search/2/search/${address}.json'`,
    {
      params: {
        key,
        limit: 100
      }
    }
  )
  const results = autocomplete.data.results.map((result) => {
    return {
      placeId: result.id,
      address: result.address,
      position: result.position
    }
  })
  return { results }
}
