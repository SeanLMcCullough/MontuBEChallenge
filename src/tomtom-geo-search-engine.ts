import axios, { AxiosInstance, type AxiosResponse } from 'axios'
import type {
  GeoSearchEngine,
  GeoSearchResults,
  GeoSearchResult,
  GeoSearchResultAddress,
  TomTomGeoSearchOptions,
  TomTomGeoSearchEngineConfiguration,
  TomTomGeoSearchResponse,
  TomTomGeoSearchResult
} from './types'

/**
 * TomTomGeoSearchEngine is a class that provides a wrapper around the TomTom API for fuzzy searching.
 */
export default class TomTomGeoSearchEngine implements GeoSearchEngine {
  private readonly _config: TomTomGeoSearchEngineConfiguration
  private readonly _api: AxiosInstance

  constructor (config: TomTomGeoSearchEngineConfiguration) {
    this._config = Object.assign(
      {
        baseUrl: 'https://api.tomtom.com',
        apiVersion: '2'
      },
      config
    )
    this._config.defaults = Object.assign(
      {
        limit: 10,
        countrySet: ['AU']
      },
      this._config.defaults
    )
    this._api = this._configureApi()
  }

  _configureApi (): AxiosInstance {
    const key = this._config.apiKey
    if (key === '') throw new Error('No API key provided')
    const api = axios.create({
      baseURL: this._config.baseUrl,
      ...(this._config.axiosConfig ?? {})
    })
    api.interceptors.request.use(config => {
      config.params.key = key
      return config
    })
    return api
  }

  /**
   * getAutoCompleteDetails provides a list of places based on a search query. The search is performed in a fuzzy manner using the TomTom API.
   * @param query the fuzzy search query
   * @param options options for performing the fuzzy search. Defaults are applied from config.defaults.
   * @returns
   */
  async getAutoCompleteDetails (query: string, options?: TomTomGeoSearchOptions): Promise<GeoSearchResults> {
    const encodedQuery = encodeURIComponent(query)
    const path = `/search/${this._config.apiVersion ?? '2'}/search/${encodedQuery}.json`
    const opts = {
      params: {
        ...this._config.defaults,
        ...options
      }
    }
    const autocomplete: AxiosResponse<TomTomGeoSearchResponse> = await this._api.get(path, opts)
    const results = mapResponse(autocomplete.data)
    return results
  }
}

function mapResponse (response: TomTomGeoSearchResponse): GeoSearchResults {
  return {
    results: response.results.map(result => mapResult(result))
  }
}

function mapResult (result: TomTomGeoSearchResult): GeoSearchResult {
  return {
    placeId: result.id,
    address: mapAddress(result.address),
    position: result.position
  }
}

function mapAddress (address: GeoSearchResultAddress): GeoSearchResultAddress {
  return {
    country: address.country ?? null,
    countryCode: address.countryCode ?? null,
    countryCodeISO3: address.countryCodeISO3 ?? null,
    countrySecondarySubdivision: address.countrySecondarySubdivision ?? null,
    countrySubdivision: address.countrySubdivision ?? null,
    countrySubdivisionCode: address.countrySubdivisionCode ?? null,
    countrySubdivisionName: address.countrySubdivisionName ?? null,
    countryTertiarySubdivision: address.countryTertiarySubdivision ?? null,
    extendedPostalCode: address.extendedPostalCode ?? null,
    freeformAddress: address.freeformAddress ?? null,
    localName: address.localName ?? null,
    municipality: address.municipality ?? null,
    municipalitySecondarySubdivision: address.municipalitySecondarySubdivision ?? null,
    municipalitySubdivision: address.municipalitySubdivision ?? null,
    neighbourhood: address.neighbourhood ?? null,
    postalCode: address.postalCode ?? null,
    streetName: address.streetName ?? null,
    streetNumber: address.streetNumber ?? null
  }
}
