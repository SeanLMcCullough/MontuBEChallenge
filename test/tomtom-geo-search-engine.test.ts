import { config } from 'dotenv'
import { describe } from '@jest/globals'
import { TomTomGeoSearchEngine } from '../src'

config()

describe('TomTomGeoSearchEngine', () => {
  it('can be constructed with a config object', () => {
    const config = {
      apiKey: process.env.TOMTOM_API_KEY
    }
    const engine = new TomTomGeoSearchEngine(config)
    expect(engine).toBeInstanceOf(TomTomGeoSearchEngine)
  })

  it('Throws an error if no api key is provided', () => {
    const config = {
      apiKey: ''
    }
    expect(() => new TomTomGeoSearchEngine(config)).toThrow()
  })
})

describe('getPlaceAutocomplete', () => {
  let engine: TomTomGeoSearchEngine

  beforeAll(() => {
    const config = {
      apiKey: process.env.TOMTOM_API_KEY
    }
    engine = new TomTomGeoSearchEngine(config)
  })

  it('returns a promise', () => {
    const res = engine.getAutoCompleteDetails('Big Pineapple')
    console.log(res)
    expect(res).toBeInstanceOf(Promise)
  })

  it('can fetch a location with a street address from the autocomplete api', async () => {
    const res = await engine.getAutoCompleteDetails('big banana')
    expect(res).toHaveProperty('results')
    expect(res.results).toBeInstanceOf(Array)
    expect(res.results.length).toBeGreaterThan(0)

    const firstRes = res.results[0]
    expect(firstRes).toHaveProperty('placeId')
    expect(firstRes).toHaveProperty('address')
    expect(firstRes).toHaveProperty('position')

    const { placeId, address, position } = firstRes
    expect(address).toHaveProperty('streetNumber')
    expect(address).toHaveProperty('countryCode')
    expect(address).toHaveProperty('country')
    expect(address).toHaveProperty('freeformAddress')
    expect(address).toHaveProperty('municipality')
    expect(address).toMatchSnapshot()
    expect(position).toHaveProperty('lat')
    expect(position).toHaveProperty('lon')
    expect(placeId).not.toBe('')
  })

  it('can fetch a location without a street address from the autocomplete api', async () => {
    const res = await engine.getAutoCompleteDetails('parliament house')
    expect(res).toHaveProperty('results')
    expect(res.results).toBeInstanceOf(Array)
    expect(res.results.length).toBeGreaterThan(0)

    const firstRes = res.results[0]
    expect(firstRes).toHaveProperty('placeId')
    expect(firstRes).toHaveProperty('address')
    expect(firstRes).toHaveProperty('position')

    const { placeId, address, position } = firstRes
    expect(address).toHaveProperty('streetNumber')
    expect(address).toHaveProperty('countryCode')
    expect(address).toHaveProperty('country')
    expect(address).toHaveProperty('freeformAddress')
    expect(address).toHaveProperty('municipality')
    expect(address).toMatchSnapshot()
    expect(position).toHaveProperty('lat')
    expect(position).toHaveProperty('lon')
    expect(placeId).not.toBe('')
  })

  it('handles a query that yields no results', async () => {
    const res = await engine.getAutoCompleteDetails(
      "You've got to be very careful if you don't know where you are going, because you might not get there."
    )
    expect(res.results).toStrictEqual([])
  })
})
