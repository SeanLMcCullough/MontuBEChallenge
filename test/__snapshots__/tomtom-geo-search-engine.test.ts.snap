// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPlaceAutocomplete can fetch a location with a street address from the autocomplete api 1`] = `
Object {
  "country": "Australia",
  "countryCode": "AU",
  "countryCodeISO3": "AUS",
  "countrySecondarySubdivision": "Lower North Coast",
  "countrySubdivision": "New South Wales",
  "countrySubdivisionCode": "NSW",
  "countrySubdivisionName": "New South Wales",
  "countryTertiarySubdivision": null,
  "extendedPostalCode": null,
  "freeformAddress": "351 Pacific Highway, Coffs Harbour, New South Wales, 2450",
  "localName": "Coffs Harbour",
  "municipality": "Coffs Harbour",
  "municipalitySecondarySubdivision": null,
  "municipalitySubdivision": null,
  "neighbourhood": null,
  "postalCode": "2450",
  "streetName": "Pacific Highway",
  "streetNumber": "351",
}
`;

exports[`getPlaceAutocomplete can fetch a location without a street address from the autocomplete api 1`] = `
Object {
  "country": "Australia",
  "countryCode": "AU",
  "countryCodeISO3": "AUS",
  "countrySecondarySubdivision": "Adelaide",
  "countrySubdivision": "South Australia",
  "countrySubdivisionCode": "SA",
  "countrySubdivisionName": "South Australia",
  "countryTertiarySubdivision": null,
  "extendedPostalCode": null,
  "freeformAddress": "North Terrace, Adelaide Central Business District, South Australia, 5000 (Rundle Mall)",
  "localName": "Adelaide Central Business District",
  "municipality": "Adelaide",
  "municipalitySecondarySubdivision": null,
  "municipalitySubdivision": "Adelaide Central Business District",
  "neighbourhood": "Rundle Mall",
  "postalCode": "5000",
  "streetName": "North Terrace",
  "streetNumber": null,
}
`;
